yupiik-batch-metrics-relay/src/main/java/io/yupiik/batch/metricsrelay/MetricsRelayEndpoint.java

```diff
@@ -0,0 +1,90 @@
+import java.util.logging.Logger;
+import java.util.stream.Stream;
+import java.util.zip.GZIPInputStream;
+
+import static io.yupiik.fusion.framework.build.api.http.HttpMatcher.PathMatching.STARTS_WITH;
+import static java.util.Optional.ofNullable;
+import static java.util.logging.Level.SEVERE;
+import static java.util.stream.Collectors.joining;
+
+@ApplicationScoped
+public class MetricsRelayEndpoint {
+
+    private final Logger logger = Logger.getLogger(MetricsRelayEndpoint.class.getName());
+
+    private final MetricsRelayStorage storage;
+
+    public MetricsRelayEndpoint(MetricsRelayStorage storage) {
+        this.storage = storage;
+    }
+
+    @HttpMatcher(path = "/relay", methods = "POST", pathMatching = STARTS_WITH)
+    public Response storeMetrics(final Request request) {
+        final String responseContentType = "text/plain";
+        final boolean dropOnPull = Boolean.parseBoolean(ofNullable(request.parameter("dropOnPull")).orElse("false"));
+        final String id = ofNullable(request.parameter("id")).orElse("");
+
+        try (final var in = read(request)) {
+            this.storage.store(id, in.lines().collect(joining("\n")).strip(), dropOnPull);
+        } catch (final IOException | ExecutionException | InterruptedException exception) {
+            logger.log(SEVERE, exception, exception::getMessage);
+            return Response.of().body("KO").header("Content-Type", responseContentType).build();
+        }
+        return Response.of().body("OK").status(201).header("Content-Type", responseContentType).build();
+    }
+
+    @HttpMatcher(path = "/relay", methods = "GET", pathMatching = STARTS_WITH)
+    public Response fetchMetrics(final Request request) {
+        final String responseContentType = "application/openmetrics-text; version=1.0.0; charset=utf-8";
+        if (this.storage.getMetrics().isEmpty()) {
+            return Response.of().header("Content-Type", responseContentType).build();
+        }
+
+        final boolean ignoreDrop = Boolean.parseBoolean(ofNullable(request.parameter("ignoreDrop")).orElse("false"));
+        final String payload = Stream.of(storage.getMetrics().toArray(new MetricsRelayStorage.Entry[0]))
+                .peek(entry -> {
+                    if (!ignoreDrop && entry.dropOnPull()) {
+                        this.storage.remove(entry); // it is a CopyOnWriteArrayList so it.remove() does not work
+                    }
+                })
+                .map(MetricsRelayStorage.Entry::content).sorted().collect(joining("\n", "", "\n# EOF\n"));
+
+        return Response.of().body(payload).status(200).header("Content-Type", responseContentType).build();
+    }
+
+    private BufferedReader read(final Request request) throws IOException, ExecutionException, InterruptedException {
+        final var encoding = request.header("content-encoding");
+        final var byteArrayIs = new ByteArrayInputStream(request.fullBody().bytes().toCompletableFuture().get());
```

Comment:
can be worth making the method returning a `CompletionStage<BufferedReader>` and the endpoint a `CompletionStage<Response>` no?